File: MusicAppDemo/MusicAppDemo/CommentTableViewController.m
//
//  CommentTableViewController.m
//  MusicAppDemo
//
//  Created by Chengzhi Jia on 15/10/18.
//  Copyright © 2015年 Chengzhi Jia. All rights reserved.
//

#import "CommentTableViewController.h"
#import "CoreDataProcess.h"
#import "LikeEntity.h"
#import "DetailEntity.h"
#import "CommentsTableViewCell.h"
#import "WriteCommentsViewController.h"

@interface CommentTableViewController ()<NSFetchedResultsControllerDelegate>


@property (strong, nonatomic) NSFetchedResultsController *fetchedResultsController;

@end

@implementation CommentTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.tableView.backgroundView = [[UIImageView alloc]initWithImage:[UIImage imageNamed:@"commentTableView"]];
    self.tableView.estimatedRowHeight = 100;
    self.tableView.rowHeight = UITableViewAutomaticDimension;
}

- (void)viewWillAppear:(BOOL)animated
{
    [self.fetchedResultsController performFetch:nil];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return self.fetchedResultsController.sections.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    id<NSFetchedResultsSectionInfo> sectionInfo = [self.fetchedResultsController sections][section];
    return [sectionInfo numberOfObjects];
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    CommentsTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"comments" forIndexPath:indexPath];
    DetailEntity  *detailEntity  = [self.fetchedResultsController objectAtIndexPath:indexPath];
    cell.detailComment.text = detailEntity.comments;
    cell.detailImage.image = [UIImage imageWithData:detailEntity.imageData];
    cell.detailImage.layer.cornerRadius = CGRectGetWidth(cell.detailImage.frame)/2.0f;
    cell.backgroundColor = [UIColor clearColor];
    return cell;
}

#pragma mark - fetch coreData

- (NSFetchRequest *)entrylistFetchRequest {
    NSFetchRequest *fetchRequest = [[NSFetchRequest alloc]initWithEntityName:@"DetailEntity"];
    fetchRequest.sortDescriptors =@[[NSSortDescriptor sortDescriptorWithKey:@"date" ascending:NO]];
    return fetchRequest;
}


- (NSFetchedResultsController *)fetchedResultsController{
    if (_fetchedResultsController != nil)
    {
        return _fetchedResultsController;
    }
    CoreDataProcess *coreDataProcess = [CoreDataProcess defaultStack];
    NSFetchRequest *fetchRequest = [self entrylistFetchRequest];
    _fetchedResultsController = [[NSFetchedResultsController alloc]initWithFetchRequest:fetchRequest managedObjectContext:coreDataProcess.managedObjectContext sectionNameKeyPath:@"section" cacheName:nil];
    
    _fetchedResultsController.delegate = self;
    
    return _fetchedResultsController;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    id<NSFetchedResultsSectionInfo> sectionInfo = [self.fetchedResultsController sections][section];
    return [sectionInfo name];
}

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return UITableViewCellEditingStyleDelete;
}

-(void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    DetailEntity *detailEntity = [self.fetchedResultsController objectAtIndexPath:indexPath];
    CoreDataProcess *core = [CoreDataProcess defaultStack];
    [core.managedObjectContext deleteObject:detailEntity];
    [core saveContext];
}

- (void)controllerWillChangeContent:(NSFetchedResultsController *)controller{
    
    [self.tableView beginUpdates];
}

-(void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)anObject atIndexPath:(NSIndexPath *)indexPath forChangeType:(NSFetchedResultsChangeType)type newIndexPath:(NSIndexPath *)newIndexPath{
    switch (type) {
        case NSFetchedResultsChangeInsert:
            [self.tableView insertRowsAtIndexPaths:@[newIndexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
            break;
            
        case NSFetchedResultsChangeDelete:
            [self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
            break;
            
        case NSFetchedResultsChangeUpdate:
            [self.tableView reloadRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
            break;
            
        default:
            break;
    }
}

-(void)controller:(NSFetchedResultsController *)controller didChangeSection:(id<NSFetchedResultsSectionInfo>)sectionInfo atIndex:(NSUInteger)sectionIndex forChangeType:(NSFetchedResultsChangeType)type{
    switch (type){
        case NSFetchedResultsChangeInsert:
            [self.tableView insertSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:UITableViewRowAnimationAutomatic];
            break;
            
        case NSFetchedResultsChangeDelete:
            [self.tableView deleteSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:UITableViewRowAnimationAutomatic];
            break;
            
        default:
            break;
    }
}

-(void)controllerDidChangeContent:(NSFetchedResultsController *)controller{
    [self.tableView endUpdates];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender{
    if ([segue.identifier isEqualToString:@"modify"])
    {
        UITableViewCell *cell = sender;
        NSIndexPath *indexPath = [self.tableView indexPathForCell:cell];
        WriteCommentsViewController *comment = segue.destinationViewController;
        comment.detailEntity = [self.fetchedResultsController objectAtIndexPath:indexPath];
    }
    
}

@end

File: MusicAppDemoInSwift/MusicAppDemoInSwift/CommentTableViewController.swift
//
//  CommentTableViewController.swift
//  MusicAppDemoInSwift
//
//  Created by Chengzhi Jia on 15/11/22.
//  Copyright © 2015年 Chengzhi Jia. All rights reserved.
//

import UIKit
import CoreData

class CommentTableViewController: UITableViewController, NSFetchedResultsControllerDelegate {
    
    lazy var fetchedController: NSFetchedResultsController = {
        var innerFetchedController = NSFetchedResultsController()
        let coreDataProcess = CoreDataProcess.defaultStack
        let fetchRequest = self.fetchRequest
        innerFetchedController = NSFetchedResultsController.init(fetchRequest: fetchRequest, managedObjectContext: coreDataProcess.managedObjectContext, sectionNameKeyPath: "section", cacheName: nil)
        innerFetchedController.delegate = self
        return innerFetchedController
    }()
    
    lazy var fetchRequest: NSFetchRequest = {
        let fetchRequest = NSFetchRequest.init(entityName: "CommentsEntity")
        fetchRequest.sortDescriptors = [NSSortDescriptor.init(key: "date", ascending: false)]
        return fetchRequest
    }()
    

    override func viewDidLoad() {
        super.viewDidLoad()
        do {
            try self.fetchedController.performFetch()
        } catch let error as NSError {
            print(error.description)
        }
        
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // MARK: - Table view data source

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return (self.fetchedController.sections?.count)!
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let sectionInfo: NSFetchedResultsSectionInfo = self.fetchedController.sections![section]
        return sectionInfo.numberOfObjects
    }

    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("thirdCell", forIndexPath: indexPath) ?? UITableViewCell.init(style: UITableViewCellStyle.Default, reuseIdentifier: "thirdCell")
        let entity = self.fetchedController.objectAtIndexPath(indexPath) as! CommentsEntity
        cell.imageView?.image = UIImage.init(data: entity.imageData!)
        cell.imageView?.layer.cornerRadius = (cell.imageView?.image?.size.width)! / 2
        cell.textLabel?.text = entity.comments
        return cell
    }
    
    override func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        let sectionInfo = self.fetchedController.sections![section]
        return sectionInfo.name
    }
    
    override func tableView(tableView: UITableView, editingStyleForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCellEditingStyle {
        return UITableViewCellEditingStyle.Delete
    }
    
    override func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath) {
        if editingStyle == .Delete {
            let comment = self.fetchedController.objectAtIndexPath(indexPath) as! CommentsEntity
            let coreData: CoreDataProcess = CoreDataProcess.defaultStack
            coreData.managedObjectContext.deleteObject(comment)
            coreData.saveContext()
        }
    }

    // mark - NSFetchedController part
    
    func controllerWillChangeContent(controller: NSFetchedResultsController) {
        self.tableView.beginUpdates()
    }
    
    func controller(controller: NSFetchedResultsController, didChangeObject anObject: AnyObject, atIndexPath indexPath: NSIndexPath?, forChangeType type: NSFetchedResultsChangeType, newIndexPath: NSIndexPath?) {
        
        switch type {
        case NSFetchedResultsChangeType.Insert:
            self.tableView.insertRowsAtIndexPaths([newIndexPath!], withRowAnimation: UITableViewRowAnimation.Automatic)
        case NSFetchedResultsChangeType.Delete:
            self.tableView.deleteRowsAtIndexPaths([indexPath!], withRowAnimation: UITableViewRowAnimation.Automatic)
        case NSFetchedResultsChangeType.Update:
            self.tableView.reloadRowsAtIndexPaths([indexPath!], withRowAnimation: UITableViewRowAnimation.Automatic)
        default:
            break
        }
    }
    
    func controller(controller: NSFetchedResultsController, didChangeSection sectionInfo: NSFetchedResultsSectionInfo, atIndex sectionIndex: Int, forChangeType type: NSFetchedResultsChangeType) {
        switch type {
        case NSFetchedResultsChangeType.Insert:
            self.tableView.insertSections(NSIndexSet.init(index: sectionIndex), withRowAnimation: UITableViewRowAnimation.Automatic)
        case NSFetchedResultsChangeType.Delete:
            self.tableView.deleteSections(NSIndexSet.init(index: sectionIndex), withRowAnimation: UITableViewRowAnimation.Automatic)
        default:
            break
        }
    }
    
    func controllerDidChangeContent(controller: NSFetchedResultsController) {
        self.tableView.endUpdates()
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "gotoUpdate" {
            if let cell = sender as? UITableViewCell {
                let indexPath: NSIndexPath = self.tableView.indexPathForCell(cell)!
                let comment = segue.destinationViewController as! WriteCommentsViewController
                comment.entity = self.fetchedController.objectAtIndexPath(indexPath) as? CommentsEntity
            }
        }
    }

}
